Value of Max_Retries - cant be passed in as var

I am trying to make max_retries value something that could be scaled by an env_var…..
say instead of 20 retries, 40 retries...

I have tried:
max_retries: !int '{service.retry_max_iters}'
max_retries: !int '{service.retry_max_iters:d}'
max_retries: !int '{service.retry_max_iters:s}'
max_retries: '{service.retry_max_iters}'
max_retries: 20                        ---------------- only one that works.
max_retries: !int "20"               -----------------edit to issue --- also works

I am getting this error for all others: 
E    - Value '<tavern.util.loader.IntToken object at 0x7f612c9030d0>' is not of type 'int'. Path: '/stages/1/max_retries'.

Is it possible to pass in a number?

Thanks for any help!

--- a/tavern/schemas/extensions.py
+++ b/tavern/schemas/extensions.py
@@ -488,3 +488,15 @@ def raise_body_error(value, rule_obj, path):
 
     msg = "The 'body' key has been replaced with 'json' in 1.0 to make it more in line with other blocks. see https://github.com/taverntesting/tavern/issues/495 for details."
     raise BadSchemaError(msg)
+
+
+def retry_variable(value, rule_obj, path):
+    """Check retry variables"""
+
+    int_variable(value, rule_obj, path)
+
+    if isinstance(value, int):
+        if value < 0:
+            raise BadSchemaError("max_retries must be greater than 0")
+
+    return True

--- a/tavern/schemas/tests.schema.yaml
+++ b/tavern/schemas/tests.schema.yaml
@@ -35,11 +35,10 @@ schema;stage:
       required: true
       unique: true
     max_retries:
-      type: int
+      type: any
+      func: retry_variable
       required: false
       unique: true
-      range:
-        min: 0
 
     skip:
       type: any

--- a/tavern/util/dict_util.py
+++ b/tavern/util/dict_util.py
@@ -107,7 +107,7 @@ def format_keys(val, variables, no_double_format=True):
             https://github.com/taverntesting/tavern/issues/431
 
     Returns:
-        dict: recursively formatted dictionary
+        str,int,list,dict: recursively formatted values
     """
     formatted = val
     box_vars = Box(variables)
@@ -127,6 +127,7 @@ def format_keys(val, variables, no_double_format=True):
         if no_double_format:
             formatted = _FormattedString(formatted)
     elif isinstance(val, TypeConvertToken):
+        logger.debug("Got type convert token '%s'", val)
         if isinstance(val, ForceIncludeToken):
             formatted = _attempt_find_include(val.value, box_vars)
         else:

--- a/tavern/util/exceptions.py
+++ b/tavern/util/exceptions.py
@@ -130,3 +130,7 @@ class InvalidExtBlockException(TavernException):
                 block
             )
         )
+
+
+class InvalidRetryException(TavernException):
+    """Invalid spec for max_retries"""

--- a/tavern/util/retry.py
+++ b/tavern/util/retry.py
@@ -3,6 +3,7 @@ import logging
 
 from . import exceptions
 from .delay import delay
+from .dict_util import format_keys
 
 logger = logging.getLogger(__name__)
 
@@ -15,7 +16,12 @@ def retry(stage, test_block_config):
         stage (dict): test stage
     """
 
-    max_retries = stage.get("max_retries", 0)
+    if "max_retries" in stage:
+        max_retries = maybe_format_max_retries(
+            stage.get("max_retries"), test_block_config
+        )
+    else:
+        max_retries = 0
 
     if max_retries == 0:
 
@@ -73,3 +79,22 @@ def retry(stage, test_block_config):
             return wrapped
 
         return retry_wrapper
+
+
+def maybe_format_max_retries(max_retries, test_block_config):
+    """Possibly handle max_retries validation"""
+
+    # Probably a format variable, or just invalid (in which case it will fail further down)
+    max_retries = format_keys(max_retries, test_block_config["variables"])
+
+    # Missing type token will mean that max_retries is still a string and will fail here
+    # Could auto convert here as well, but keep it consistent and just fail
+    if not isinstance(max_retries, int):
+        raise exceptions.InvalidRetryException(
+            "Invalid type for max_retries - was {}".format(type(max_retries))
+        )
+
+    if max_retries < 0:
+        raise exceptions.InvalidRetryException("max_retries must be greater than 0")
+
+    return max_retries

--- a/tests/integration/global_cfg.yaml
+++ b/tests/integration/global_cfg.yaml
@@ -6,3 +6,6 @@ variables:
   global_host: http://localhost:5003
   global_test_string_1: abc
   global_test_string_2: "{tavern.env_vars.SECOND_URL_PART}"
+
+  retry_max: 4
+  negative_int: -2

--- a/tests/integration/test_retry.tavern.yaml
+++ b/tests/integration/test_retry.tavern.yaml
@@ -38,6 +38,84 @@ stages:
 
 ---
 
+test_name: Format max retry variable correctly
+
+includes:
+  - !include common.yaml
+
+stages:
+  - name: polling
+    max_retries: !int '{retry_max}'
+    request:
+      url: "{host}/poll"
+      method: GET
+    response:
+      status_code: 200
+      json:
+        status: ready
+
+---
+
+test_name: Format max retry variable fails if not using type token
+
+includes:
+  - !include common.yaml
+
+_xfail: verify
+
+stages:
+  - name: polling
+    max_retries: '{retry_max}'
+    request:
+      url: "{host}/poll"
+      method: GET
+    response:
+      status_code: 200
+      json:
+        status: ready
+
+---
+
+test_name: Format max retry variable fails if invalid value
+
+includes:
+  - !include common.yaml
+
+_xfail: run
+
+stages:
+  - name: polling
+    max_retries: !int '{negative_int}'
+    request:
+      url: "{host}/poll"
+      method: GET
+    response:
+      status_code: 200
+      json:
+        status: ready
+
+---
+
+test_name: Format max retry variable fails if using wrong type token
+
+includes:
+  - !include common.yaml
+
+_xfail: verify
+
+stages:
+  - name: polling
+    max_retries: !float '{retry_max}'
+    request:
+      url: "{host}/poll"
+      method: GET
+    response:
+      status_code: 200
+      json:
+        status: ready
+
+---
+
 test_name: Setting max_retries to less than 0 should fail
 
 includes:

